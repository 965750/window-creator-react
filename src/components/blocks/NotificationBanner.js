import React, { Component } from 'react'
import styles from './NotificationBanner.module.scss'
import { connect } from 'react-redux'
import { clearNotification } from '../../store/actions/themeActions'

class NotificationBanner extends Component {
    state = {
        show: false,
    }

    onCloseBanner = time => {
        setTimeout(() => {
            this.props.clearNotification()

            this.setState({
                show: false,
            })
        }, time)
    }

    componentDidUpdate() {
        if (this.props.notification && !this.state.show) {
            this.setState({
                show: true,
            })

            this.onCloseBanner(7000)
        }
    }

    render() {
        if (this.state.show) {
            return (
                <div
                    className={`${styles.banner} bg-error fixed w-full z-30 top-0 text-white`}
                >
                    <p className={`${styles.banner__text} text-center`}>
                        {this.props.notification}
                    </p>
                    <span
                        onClick={() => this.onCloseBanner(0)}
                        className={`${styles.banner__close} absolute right-30 top-half`}
                    >
                        &#10005;
                    </span>
                </div>
            )
        } else {
            return null
        }
    }
}

const mapDispatchToProps = dispatch => {
    return {
        clearNotification: () => dispatch(clearNotification()),
    }
}

const mapStateToProps = state => {
    console.log(state, 11111)
    return {
        notification: state.theme.notification,
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(NotificationBanner)
