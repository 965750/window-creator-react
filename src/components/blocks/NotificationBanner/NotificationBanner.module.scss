.banner {
    height: 80px;

    &__text {
        width: 70%
    }

    &__close {
        transform: translate(0, -50%)
    }
}