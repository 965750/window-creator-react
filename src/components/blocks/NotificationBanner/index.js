import { clearNotification } from '../../../store/actions/themeActions'
import { connect } from 'react-redux'
import { FormattedMessage } from 'react-intl'

import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './NotificationBanner.module.scss'

class NotificationBanner extends Component {
  state = {
    show: false,
  }

  componentDidUpdate() {
    this.startNotification()
  }

  startNotification = () => {
    if (this.props.notification && !this.state.show) {
      this.setState({
        show: true,
      })

      this.onCloseBanner(5000)
    }
  }

  onCloseBanner = (time) => {
    setTimeout(() => {
      this.props.clearNotification()

      this.setState({
        show: false,
      })
    }, time)
  }

  render() {
    if (this.state.show && this.props.notification) {
      return (
        <div
          className={`${styles.banner} bg-${this.props.notification.type} flex items-center fixed w-full z-50 top-0 text-white`}
        >
          <p className={`${styles.banner__text} mx-auto text-center`}>
            <FormattedMessage
              id={this.props.notification.text}
              defaultMessage={this.props.notification.text}
            />
          </p>
          <button
            data-testid="closeNotification"
            onClick={() => this.onCloseBanner(0)}
            className={`${styles.banner__close} absolute right-30 cursor-pointer top-half`}
          >
            &#10005;
          </button>
        </div>
      )
    }
    return null
  }
}

const mapDispatchToProps = (dispatch) => ({
  clearNotification: () => dispatch(clearNotification()),
})

const mapStateToProps = (state) => ({
  notification: state.theme.notification,
})

NotificationBanner.propTypes = {
  clearNotification: PropTypes.func,
  notification: PropTypes.shape({
    type: PropTypes.string,
    text: PropTypes.string,
  }),
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(NotificationBanner)
