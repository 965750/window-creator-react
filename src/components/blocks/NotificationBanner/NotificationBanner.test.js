import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react'

import NotificationBanner from './index'
import messages from '../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
    notification: {
      text: 'mockNotificationText',
      type: 'mockType'
    }
  },
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    case 'MOCK_UPDATE_COMPONENT':
      return {
        ...state,
        theme: {
          notification: {
            text: 'mockNotificationText2',
            type: 'mockType2'
          }
        }
      }
    case 'CLEAR_NOTIFICATION':
      return {
        ...state,
        theme: {
          notification: null
        }
      }
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('NotificationBanner', () => {
  let wrapper
  const initialProps = {}

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <NotificationBanner {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match empty snapshot initiali', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should match full notification snapshot', () => {
    setUp()

    store.dispatch({
      type: 'MOCK_UPDATE_COMPONENT',
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should show up then hide notification after close button was clicked', () => {
    jest.useFakeTimers()
    setUp()

    store.dispatch({
      type: 'MOCK_UPDATE_COMPONENT',
    })

    const closeBtn = wrapper.getByTestId("closeNotification")
    fireEvent.click(closeBtn)

    jest.runOnlyPendingTimers();

    expect(wrapper.asFragment()).toMatchSnapshot()
  })
})
