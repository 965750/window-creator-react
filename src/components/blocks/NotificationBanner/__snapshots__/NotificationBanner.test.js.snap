// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBanner should match empty snapshot initiali 1`] = `<DocumentFragment />`;

exports[`NotificationBanner should match full notification snapshot 1`] = `
<DocumentFragment>
  <div
    class="banner bg-mockType2 flex items-center fixed w-full z-50 top-0 text-white"
  >
    <p
      class="banner__text mx-auto text-center"
    >
      mockNotificationText2
    </p>
    <button
      class="banner__close absolute right-30 cursor-pointer top-half"
      data-testid="closeNotification"
    >
      âœ•
    </button>
  </div>
</DocumentFragment>
`;

exports[`NotificationBanner should show up then hide notification after close button was clicked 1`] = `<DocumentFragment />`;
